### ISSUE TITLE
Improvement: maybe use Erlang binary syntax

### ISSUE THREAD

**Issue by karlsson** [UncleGrumpy/atomvm_dotstar#2]
Hi,
I am just thinking that the Erlang binary syntax could be handy and save you from some bsl/bor instructions in dotstar.erl:
```
write_dot(Red, Green, Blue, Illumination) ->
    %% fast and dirty way to convert 0-100 Illumination into 5-bit Brightness.
    Bright = (Illumination * 31) div 97,
    Dot_data = <<?SPI_START_FRAME:32, ?SPI_START_BITS:3, Bright:5,
               Blue:8, Green:8, Red:8, ?SPI_END_FRAME:32>>,
    Transaction = #{command => 0, address => 0, write_data => Dot_data, write_bits => 96, read_bits => 0},
    spi:write(erlang:whereis(dotstar), ?LED_DEVICE_NAME, Transaction).
```
`-define(SPI_START_BITS, 2#111).`

And case clause returns values:
```
write_dot_hsv(H, S, V) ->
    I = (V * 31) div 97,
    RGB_max = (V * 255) div 100,
    RGB_min = (RGB_max * (100 - S)) div 100,
    Diff = H rem 60,
    RGB_adj = ((RGB_max - RGB_min) * Diff) div 60,
    Sextant = H div 60,
    {R,G,B} =
        case Sextant of
            0 -> {RGB_max, RGB_min + RGB_adj, RGB_min};
            1 -> {RGB_max - RGB_adj, RGB_max, RGB_min};
            2 -> {RGB_min, RGB_max, RGB_min + RGB_adj};
            3 -> {RGB_min, RGB_max - RGB_adj, RGB_max};
            4 -> {RGB_min + RGB_adj, RGB_min, RGB_max};
            _ -> {RGB_max, RGB_min, RGB_max - RGB_adj}
        end,
    write_dot(R, G, B, I).
```

**Comment by karlsson** [comment]
If you extend for more dots maybe a list with one tuple for every dot?
You can use the binary syntax in Accumulators too.
I assume that the first byte start with 2#111 for the first dot and then 2#000 for the rest (the brightness bytes).
```
write_dot(Red, Green, Blue, Illumination) ->
    write_dots([{Red, Green, Blue, Illumination}]).

write_dots(DotList) ->
    DS = dot_stream(DotList),
    DotData = <<?SPI_START_FRAME:32, DS/binary, ?SPI_END_FRAME:32>>,
    Transaction = #{write_data => DotData},
    spi:write(erlang:whereis(dotstar), ?LED_DEVICE_NAME, Transaction).

%% Private functions
dot_stream(DotList) ->
    dot_stream(DotList, <<>>).

dot_stream([], Acc) ->
    Acc;
dot_stream([{R,G,B,I}|T], <<>>) -> %% First dot
    Bright = bright(I),
    dot_stream(T, <<?SPI_START_BITS:3, Bright:5, B:8, G:8, R:8>>);
dot_stream([{R,G,B,I}|T], Acc) ->
    Bright = bright(I),
    dot_stream(T, << Acc/binary, 0:3, Bright:5, B:8, G:8, R:8>>).

%% fast and dirty way to convert 0-100 Illumination into 5-bit Brightness.
bright(Illumination) ->
    (Illumination * 31) div 97.
```

**Comment by UncleGrumpy** [comment]
Absolutely! I just realized I wrote this driver amidst mostly writing C code for the VM itself, and was paying more attention to the datasheet than I was thinking in Erlang 🤣 . I'm also wondering if I should change the repository name to something more generic like atomvm_led_strip, since other SPI LEDs can just as easily be added to this driver. I had started it because the AP102 (A.K.A. DotStar) is different enough that the `atomvm_neopixel` (WS2812) driver would be an impractical place to add support... Or do you think it would be better to have one lower level driver per repository, and have a separate `atomvm_led_strip` library for a higher level way to control strips, with patterns and effects, which just depends on the appropriate lower level drivers?

**Comment by UncleGrumpy** [comment]
> If you extend for more dots maybe a list with one tuple for every dot?

I was thinking of accepting a list of `{Dot {R, G, B}}`, where there Dot would replace the state of that dot in the state held by the gen_server, where the `State` would be the current binary to transmit for the entire string... The AP102 work a bit differently than WS2812, aside from requiring an SPI clock signal, the way to set the color of a particular LED is to transmit the binary data describing all of the LEDs up to and including the last one that needs the color to be changed, followed by the termination frame.  There is no rate limit like the WS2812, you can continuously blast bits at any rate the MCU can support.

**Comment by UncleGrumpy** [comment]
It might be a lot more efficient to use the tuple list, as you suggested, for state now that I think about it. Holding and processing one large data chunk would probably be very inefficient on the esp32.

**Comment by UncleGrumpy** [comment]
> I think I read that ETS was coming into the AtomVM?

Indeed. Basic support for ets was merged a few months ago:
https://github.com/atomvm/AtomVM/pull/1054

I have not worked with it yet, in fact I am not familiar with ets… I started learning Erlang when I was first introduced to AtomVM, it has been a good place to start learning the basics, but I am still much less familiar with all of the extra bells and whistles that come with OTP.

**Comment by karlsson** [comment]
I noticed that Illumination is calculated as:
`I = (V * 31) div 97,` in `dotstar:write_dot_hsv/3` line 91
and then Bright is calculated as `Bright = (Illumination * 31) div 97,`  (line 68 in write_dot/4).
It looks as the calculation is done twice?

**Comment by UncleGrumpy** [comment]
Indeed. I think this explains a bug I was observing in the example I provided... That is definitely a mistake!

**Comment by UncleGrumpy** [comment]
I haven't started on multiple LED support yet, I realized I have the strip, and a power supply... but I don't have a AC power cord to sacrifice for the power supply, I need to find the time to stop coding for long enough to go into town and get a power cord 😂.

**Comment by karlsson** [comment]
Always these hardware vs. software conflicts :grin: .
I will see if I can get my hands on a neopixel type strip this weekend. I have the first "dot" on the ESP32 board but it would be nice to see multiple LEDs.
From the atomvm_neopixel application and your code it looks pretty straightforward to add SPI neopixel support in the led_strip application later. It seems as just to generate the Erlang binaries in a slightly different way. Neopixel (WS2812) seems to use only RGB values and no brightness part, but otherwise very much the same.

**Comment by UncleGrumpy** [comment]
I was noticing the subtle differences, but I think it would be easy to write a higher level led_strip.erl (which will become the main application) that can be configured to use any supported SPI LED type, once they are supported with an appropriate hardware module.

**Comment by karlsson** [comment]
I started to play around (again) with this a bit more and implemented the different LED types encoding as callback modules implementing an Erlang behaviour `avm_ls_strip` . Since I only have a Neopixel strip myself I have not been able to verify the other types.
Github: [avm_ls](https://github.com/karlsson/avm_ls)
I also tested if I could compile Erlang, Elixir and Gleam code into one flash image as well :smile: .
